src/backend/distributed/executor/intermediate_result_encoder.c

```diff
@@ -0,0 +1,289 @@
+/*-------------------------------------------------------------------------
+ *
+ * intermediate_result_encoder.c
+ *   Functions for encoding and decoding intermediate results.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "postgres.h"
+#include "funcapi.h"
+#include "libpq-fe.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+
+#include "catalog/pg_enum.h"
+#include "commands/copy.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/connection_management.h"
+#include "distributed/intermediate_results.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_client_executor.h"
+#include "distributed/multi_executor.h"
+#include "distributed/remote_commands.h"
+#include "distributed/transmit.h"
+#include "distributed/transaction_identifier.h"
+#include "distributed/tuplestore.h"
+#include "distributed/version_compat.h"
+#include "distributed/worker_protocol.h"
+#include "mb/pg_wchar.h"
+#include "nodes/makefuncs.h"
+#include "nodes/parsenodes.h"
+#include "nodes/primnodes.h"
+#include "storage/fd.h"
+#include "tcop/tcopprot.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/memutils.h"
+#include "utils/syscache.h"
+
+
+/* internal state of intermediate result file encoder */
+struct IntermediateResultEncoder
+{
+	/*
+	 * Result of encoding is accumulated in outputBuffer, until it is flushed.
+	 * The flush is done when the data is returned as result of
+	 * IntermediateResultEncoderReceive() and IntermediateResultEncoderDone()
+	 * functions.
+	 */
+	StringInfo outputBuffer;
```

Comment:
Since the caller now supplies the StringInfo, I'm kind of wondering whether it would make sense to pass it to every function call (and set it in the copyOutState). That way it's more obvious that it's appending to the caller's buffer, and it gives the caller some more flexibility (e.g. append to two different buffers). That said, I'm also fine with the current approach.